### PR TITLE
Use bazel managed node_modules for downstream angular from source build support

### PR DESCRIPTION
**Pull request by gregmagolan**
Based on #24881 which should land first.

### CODE REVIEW

File: integration/bazel/package.json [hunk 1]
```diff
@@ -21,7 +21,8 @@
     "concurrently": "3.5.1",
     "http-server": "0.11.1",
     "protractor": "file:../../node_modules/protractor",
-    "typescript": "file:../../node_modules/typescript"
+    "tsutils": "2.20.0",
+    "typescript": "2.7.2"
```
Comment (gregmagolan) [inline integration/bazel/package.json]: Still need to figure out why having typescript 2.8.4 causes a typescript/tsutils skew issue here

File: package.json [hunk 2]
```diff
@@ -21,7 +21,8 @@
     "prebuildifier": "bazel build --noshow_progress @com_github_bazelbuild_buildtools//buildifier",
     "buildifier": "find . -type f \\( -name BUILD -or -name BUILD.bazel \\) ! -path \"*/node_modules/*\" | xargs $(bazel info bazel-bin)/external/com_github_bazelbuild_buildtools/buildifier/*/buildifier",
     "preinstall": "node tools/yarn/check-yarn.js",
-    "postinstall": "yarn update-webdriver && node ./tools/postinstall-patches.js",
+    "postinstall": "yarn update-webdriver && node ./tools/postinstall-patches.js && yarn patch-types",
+    "patch-types": "mkdir node_modules/@types/zone.js && cp node_modules/zone.js/dist/zone.js.d.ts node_modules/@types/zone.js/index.d.ts",
```
Comment (gregmagolan) [inline package.json]: Work-around for being able to use `/// <reference types="zone.js" />` in `modules/types.d.ts` which is needed for the both the angular build & a downstream build from source to work

File: packages/bazel/src/ngc-wrapped/BUILD.bazel [hunk 3]
```diff
@@ -33,7 +33,7 @@ nodejs_binary(
         "@build_bazel_rules_typescript//internal:worker_protocol.proto",
     ],
     entry_point = "angular/packages/bazel/src/ngc-wrapped/index.js",
-    node_modules = "@build_bazel_rules_typescript_tsc_wrapped_deps//:node_modules",
+    node_modules_list = ["@build_bazel_rules_typescript_tsc_wrapped_deps//:node_modules", "@//:node_modules"],
```
Comment (gregmagolan) [inline packages/bazel/src/ngc-wrapped/BUILD.bazel]: See https://github.com/bazelbuild/rules_nodejs/pull/240

File: tools/defaults.bzl [hunk 4]
```diff
@@ -37,17 +38,17 @@ PKG_GROUP_REPLACEMENTS = {
     ]""" % ",\n      ".join(["\"%s\"" % s for s in ANGULAR_SCOPED_PACKAGES])
 }
 
-def ts_library(tsconfig = None, **kwargs):
+def ts_library(tsconfig = None, node_modules = DEFAULT_NODE_MODULES, **kwargs):
```
Comment (gregmagolan) [inline tools/defaults.bzl]: These defaults are needed for the downstream angular from source build

File: WORKSPACE [hunk 5]
```diff
@@ -6,9 +6,9 @@ workspace(name = "angular")
 
 http_archive(
     name = "build_bazel_rules_nodejs",
-    url = "https://github.com/bazelbuild/rules_nodejs/archive/0.10.1.zip",
-    strip_prefix = "rules_nodejs-0.10.1",
-    sha256 = "634206524d90dc03c52392fa3f19a16637d2bcf154910436fe1d669a0d9d7b9c",
+    url = "https://github.com/gregmagolan/rules_nodejs/archive/43cdff28de0e8b127e7c5fec2cfc4793a2c630a3.zip",
```
Comment (gregmagolan) [inline WORKSPACE]: Update once https://github.com/bazelbuild/rules_nodejs/pull/245 lands

Comment (gregmagolan) [reply WORKSPACE]: Done

File: package.json [hunk 6]
```diff
@@ -21,7 +21,8 @@
     "prebuildifier": "bazel build --noshow_progress @com_github_bazelbuild_buildtools//buildifier",
     "buildifier": "find . -type f \\( -name BUILD -or -name BUILD.bazel \\) ! -path \"*/node_modules/*\" | xargs $(bazel info bazel-bin)/external/com_github_bazelbuild_buildtools/buildifier/*/buildifier",
     "preinstall": "node tools/yarn/check-yarn.js",
-    "postinstall": "yarn update-webdriver && node ./tools/postinstall-patches.js",
+    "postinstall": "yarn update-webdriver && node ./tools/postinstall-patches.js && yarn patch-types",
+    "patch-types": "mkdir -p node_modules/@types/zone.js && cp -f node_modules/zone.js/dist/zone.js.d.ts node_modules/@types/zone.js/index.d.ts",
```
Comment (alexeagle) [inline package.json]: TODO/issue filed to fix the hack? users will need this hack too

**Review by alexeagle — APPROVED** [review]

File: index.bzl [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# Copyright Google Inc. All Rights Reserved.
+#
+# Use of this source code is governed by an MIT-style license that can be
+# found in the LICENSE file at https://angular.io/license
+""" Public API surface is re-exported here.
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply index.bzl]: roger

File: package.json (hunk 6 above)
Comment (gregmagolan) [reply package.json]: I'll file an issue for future reference. User's won't need to do anything as this hack will occur in the postinstall step of @angular_deps. Its only needed if you need the zone types to be in the build without an zonejs require as is done in the angular build: `/// <reference types="zone.js" />`

**Review by IgorMinar — APPROVED** [review]

File: index.bzl [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+This API is exported for user's building angular from source in downstream
```
Comment (petebacondarwin) [inline index.bzl]: nit: user's -> users

File: packages/bazel/src/ngc-wrapped/index.ts [hunk 9]
```diff
@@ -207,6 +207,12 @@ export function compile({allowNonHermeticReads, allDepsCompiledWithBazel = true,
     if (fileName ===
         path.join(compilerOpts.baseUrl, bazelOpts.package, compilerOpts.flatModuleOutFile + '.ts'))
       return true;
+    // Also handle the case when angular is build from source as an external repository
```
Comment (petebacondarwin) [inline packages/bazel/src/ngc-wrapped/index.ts]: nit: build -> built

**Pull request closed without merging**